beacon_node/eth2_libp2p/src/types/pubsub.rs

```diff
@@ -55,21 +29,65 @@ pub enum PubsubMessage<T: EthSpec> {
     AttesterSlashing(Box<AttesterSlashing<T>>),
 }
 
-fn decompress_snappy(data: &[u8]) -> Result<Vec<u8>, String> {
-    // Exit early if uncompressed data is > GOSSIP_MAX_SIZE
-    match decompress_len(data) {
-        Ok(n) if n > GOSSIP_MAX_SIZE => {
-            return Err("ssz_snappy decoded data > GOSSIP_MAX_SIZE".into());
+// Implements the `DataTransform` trait of gossipsub to employ snappy compression
+pub struct SnappyTransform {
+    /// Sets the maximum size we allow gossipsub messages to decompress to.
+    max_size_per_message: usize,
+}
+
+impl SnappyTransform {
+    pub fn new(max_size_per_message: usize) -> Self {
+        SnappyTransform {
+            max_size_per_message,
         }
-        Ok(_) => {}
-        Err(e) => {
-            return Err(format!("{}", e));
+    }
+}
+
+impl DataTransform for SnappyTransform {
+    // Provides the snappy decompression from RawGossipsubMessages
+    fn inbound_transform(
+        &self,
+        raw_message: RawGossipsubMessage,
+    ) -> Result<GossipsubMessage, std::io::Error> {
+        // check the length of the raw bytes
+        let len = decompress_len(&raw_message.data)?;
+        if len > self.max_size_per_message {
+            return Err(Error::new(
+                ErrorKind::InvalidData,
+                "ssz_snappy decoded data > GOSSIP_MAX_SIZE",
+            ));
         }
-    };
-    let mut decoder = Decoder::new();
-    match decoder.decompress_vec(data) {
-        Ok(decompressed_data) => Ok(decompressed_data),
-        Err(e) => Err(format!("{}", e)),
+
+        let mut decoder = Decoder::new();
+        let decompressed_data = decoder.decompress_vec(&raw_message.data)?;
+
+        // Build the GossipsubMessage struct
+        Ok(GossipsubMessage {
+            source: raw_message.source,
+            data: decompressed_data,
+            sequence_number: raw_message.sequence_number,
+            topic: raw_message.topic,
+        })
+    }
+
+    /// Provides the snappy compression logic to gossipsub.
+    fn outbound_transform(
+        &self,
+        _topic: &TopicHash,
+        data: Vec<u8>,
+    ) -> Result<Vec<u8>, std::io::Error> {
+        // Currently we are not employing topic-based compression. Everything is expected to be
+        // snappy compressed.
+        let mut encoder = Encoder::new();
+        let compressed_data = encoder.compress_vec(&data)?;
+
+        if compressed_data.len() > self.max_size_per_message {
```

Comment:
I would have thought we need to check the length of `data` instead of `compressed_data`.

Reply:
Yeah, this is outbound, we shouldn't hit it, but would be stricter to put it on data, have updated